sdk/core/azure-core-experimental/azure/core/experimental/transport/_urllib3.py

```diff
@@ -0,0 +1,379 @@
+            request=request,
+            internal_response=internal_response,
+            status_code=kwargs.pop("status_code", internal_response.status),
+            headers=headers,
+            reason=kwargs.pop("reason", internal_response.reason),
+            content_type=kwargs.pop("content-type", headers.get("content-type")),
+            stream_download_generator=kwargs.pop("stream_download_generator", Urllib3StreamDownloadGenerator),
+            **kwargs,
+        )
+
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        # Remove the unpicklable entries.
+        state["_internal_response"] = None  # urllib3 response are not pickable (see headers comments)
+        return state
+
+    def close(self) -> None:
+        super().close()
+        self._internal_response.release_conn()
+
+
+class Urllib3Transport(HttpTransport[RestHttpRequest, RestHttpResponse]):
+    """Implements a basic urllib3 HTTP sender.
+
+    Missing features:
+
+     - Support for per-request proxy configuration override. Proxy configuration is currently
+       only supported at client/transport construction.
+
+    :keyword pool: A preconfigured urllib3 PoolManager.
+    :paramtype pool: ~urllib3.PoolManager or None
+    """
+
+    def __init__(self, *, pool: Optional[urllib3.PoolManager] = None, **kwargs: Any) -> None:
+        self._pool: Optional[urllib3.PoolManager] = pool
+        self._pool_owner: bool = kwargs.get("pool_owner", True)
+        self._config = ConnectionConfiguration(**kwargs)
+        self._pool_cls: Callable[..., urllib3.PoolManager] = urllib3.PoolManager
+        if "proxies" in kwargs:
+            proxies = kwargs.pop("proxies")
+            if len(proxies) > 1:
+                raise ValueError("Only a single proxy url is supported for urllib3.")
+            proxy_url = list(proxies.values())[0]
+            self._pool_cls = functools.partial(urllib3.ProxyManager, proxy_url)
+
+        # See https://github.com/Azure/azure-sdk-for-python/issues/25640 to understand why we track this
+        self._has_been_opened = False
+
+    def _cert_verify(self, kwargs: Dict[str, Any]) -> None:
+        """Update the request kwargs to configure the SSL certificate.
+
+        :param dict[str, Any] kwargs: The request context keyword args.
+        """
+        verify: Union[bool, str] = kwargs.pop("connection_verify", self._config.verify)
+        cert_kwargs: Dict[str, Optional[str]] = {}
+        if verify is False:
+            # We ignore verify=True as "CERT_REQUIRED" is the default for HTTPS.
+            cert_kwargs["cert_reqs"] = "CERT_NONE"
+        elif os.path.isdir(verify):
+            cert_kwargs["ca_cert_dir"] = cast(str, verify)
+        elif isinstance(verify, str):
+            cert_kwargs["ca_certs"] = verify
+
+        cert: Optional[str] = kwargs.pop("connection_cert", self._config.cert)
+        if cert:
+            if not isinstance(cert, str):
+                cert_kwargs["cert_file"] = cert[0]
```

Comment:
Incorrect typing? `cert` is truth:y, so it is not `None`. But the type is `str | None`. So why is this check needed unless typing on L247 is incorrect?

Reply:
Yeah that's an incorrect type hint. Fixed~